# Analysis of stack overflow developer survey

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('data/survey_results_public.csv', index_col='Respondent')
schema_df = pd.read_csv('data/survey_results_schema.csv', index_col='Column')

In [3]:
df.head()

MainBranch Hobbyist   Age  \
Respondent                                                                     
1                              I am a developer by profession      Yes   NaN   
2                              I am a developer by profession       No   NaN   
3                                 I code primarily as a hobby      Yes   NaN   
4                              I am a developer by profession      Yes  25.0   
5           I used to be a developer by profession, but no...      Yes  31.0   

           Age1stCode CompFreq  CompTotal  ConvertedComp             Country  \
Respondent                                                                     
1                  13  Monthly        NaN            NaN             Germany   
2                  19      NaN        NaN            NaN      United Kingdom   
3                  15      NaN        NaN            NaN  Russian Federation   
4                  18      NaN        NaN            NaN             Albania   
5                  16      NaN        NaN            NaN       United States   

              CurrencyDesc CurrencySymbol  ...                  SurveyEase  \
Respondent                                 ...                               
1            European Euro            EUR  ...  Neither easy nor difficult   
2           Pound sterling            GBP  ...                         NaN   
3                      NaN            NaN  ...  Neither easy nor difficult   
4             Albanian lek            ALL  ...                         NaN   
5                      NaN            NaN  ...                        Easy   

                     SurveyLength Trans  \
Respondent                                
1           Appropriate in length    No   
2                             NaN   NaN   
3           Appropriate in length   NaN   
4                             NaN    No   
5                       Too short    No   

                                               UndergradMajor  \
Respondent                                                      
1           Computer science, computer engineering, or sof...   
2           Computer science, computer engineering, or sof...   
3                                                         NaN   
4           Computer science, computer engineering, or sof...   
5           Computer science, computer engineering, or sof...   

           WebframeDesireNextYear    WebframeWorkedWith  \
Respondent                                                
1                    ASP.NET Core  ASP.NET;ASP.NET Core   
2                             NaN                   NaN   
3                             NaN                   NaN   
4                             NaN                   NaN   
5            Django;Ruby on Rails         Ruby on Rails   

                                       WelcomeChange WorkWeekHrs YearsCode  \
Respondent                                                                   
1            Just as welcome now as I felt last year        50.0        36   
2           Somewhat more welcome now than last year         NaN         7   
3           Somewhat more welcome now than last year         NaN         4   
4           Somewhat less welcome now than last year        40.0         7   
5            Just as welcome now as I felt last year         NaN        15   

           YearsCodePro  
Respondent               
1                    27  
2                     4  
3                   NaN  
4                     4  
5                     8  

[5 rows x 60 columns]

In [4]:
df.shape

(64461, 60)

In [5]:
df.columns

Index(['MainBranch', 'Hobbyist', 'Age', 'Age1stCode', 'CompFreq', 'CompTotal',
       'ConvertedComp', 'Country', 'CurrencyDesc', 'CurrencySymbol',
       'DatabaseDesireNextYear', 'DatabaseWorkedWith', 'DevType', 'EdLevel',
       'Employment', 'Ethnicity', 'Gender', 'JobFactors', 'JobSat', 'JobSeek',
       'LanguageDesireNextYear', 'LanguageWorkedWith',
       'MiscTechDesireNextYear', 'MiscTechWorkedWith',
       'NEWCollabToolsDesireNextYear', 'NEWCollabToolsWorkedWith', 'NEWDevOps',
       'NEWDevOpsImpt', 'NEWEdImpt', 'NEWJobHunt', 'NEWJobHuntResearch',
       'NEWLearn', 'NEWOffTopic', 'NEWOnboardGood', 'NEWOtherComms',
       'NEWOvertime', 'NEWPurchaseResearch', 'NEWPurpleLink', 'NEWSOSites',
       'NEWStuck', 'OpSys', 'OrgSize', 'PlatformDesireNextYear',
       'PlatformWorkedWith', 'PurchaseWhat', 'Sexuality', 'SOAccount',
       'SOComm', 'SOPartFreq', 'SOVisitFreq', 'SurveyEase', 'SurveyLength',
       'Trans', 'UndergradMajor', 'WebframeDesireNextYear',
       'Webframe

In [6]:
df['ConvertedComp'].head(20)

Respondent
1           NaN
2           NaN
3           NaN
4           NaN
5           NaN
6           NaN
7           NaN
8      116000.0
9           NaN
10      32315.0
11      40070.0
12      14268.0
13      38916.0
14      66000.0
15          NaN
16     108576.0
17      79000.0
18    1260000.0
19      83400.0
20          NaN
Name: ConvertedComp, dtype: float64

In [7]:
#out of 64461 only 34756 answerd the salary question
df['ConvertedComp'].count()

34756

In [8]:
#get median of convertedcomp
df['ConvertedComp'].median()

54049.0

In [9]:
df.median()

Age                 29.0
CompTotal        63000.0
ConvertedComp    54049.0
WorkWeekHrs         40.0
dtype: float64

In [10]:
df.describe().T

count           mean           std  min      25%      50%  \
Age            45446.0   3.083411e+01  9.585392e+00  1.0     24.0     29.0   
CompTotal      34826.0  3.190464e+242           inf  0.0  20000.0  63000.0   
ConvertedComp  34756.0   1.037561e+05  2.268853e+05  0.0  24648.0  54049.0   
WorkWeekHrs    41151.0   4.078217e+01  1.781638e+01  1.0     40.0     40.0   

                    75%            max  
Age                35.0   2.790000e+02  
CompTotal      125000.0  1.111111e+247  
ConvertedComp   95000.0   2.000000e+06  
WorkWeekHrs        44.0   4.750000e+02

In [11]:
df['Hobbyist']

Respondent
1        Yes
2         No
3        Yes
4        Yes
5        Yes
        ... 
64858    Yes
64867    Yes
64898    Yes
64925    Yes
65112    Yes
Name: Hobbyist, Length: 64461, dtype: object

In [12]:
df['Hobbyist'].value_counts()

Yes    50388
No     14028
Name: Hobbyist, dtype: int64

### Analyze the column/feature platformworkedwith 

In [13]:
df['PlatformWorkedWith']

Respondent
1                                           Windows
2                                               iOS
3                                               NaN
4                                               NaN
5                    AWS;Docker;Linux;MacOS;Windows
                            ...                    
64858                                           NaN
64867                                           NaN
64898                                           NaN
64925                                 Linux;Windows
65112    Android;Arduino;Linux;Raspberry Pi;Windows
Name: PlatformWorkedWith, Length: 64461, dtype: object

In [14]:
#to find exact questions that was asked for flatformworkedwith 
schema_df.loc['PlatformWorkedWith']

QuestionText    Which platforms have you done extensive develo...
Name: PlatformWorkedWith, dtype: object

In [15]:
#to find the counts of those unique platform
df['PlatformWorkedWith'].value_counts()

Windows                                                                          4770
Linux;Windows                                                                    2461
Linux                                                                            2112
Android                                                                          1201
Microsoft Azure;Windows                                                           996
                                                                                 ... 
Docker;Heroku;Linux;MacOS;Raspberry Pi;Slack Apps and Integrations                  1
Android;Heroku;IBM Cloud or Watson;Linux;MacOS;Raspberry Pi;Windows;WordPress       1
AWS;Docker;Google Cloud Platform;Microsoft Azure;Raspberry Pi;Windows               1
Android;Docker;MacOS;Microsoft Azure                                                1
AWS;Docker;Kubernetes;MacOS;Raspberry Pi;Windows                                    1
Name: PlatformWorkedWith, Length: 6287, dtype: int64

In [16]:
#to see the percentage instead of counts use normalize arg
df['PlatformWorkedWith'].value_counts(normalize=True)

Windows                                                                          0.088591
Linux;Windows                                                                    0.045707
Linux                                                                            0.039225
Android                                                                          0.022306
Microsoft Azure;Windows                                                          0.018498
                                                                                   ...   
Docker;Heroku;Linux;MacOS;Raspberry Pi;Slack Apps and Integrations               0.000019
Android;Heroku;IBM Cloud or Watson;Linux;MacOS;Raspberry Pi;Windows;WordPress    0.000019
AWS;Docker;Google Cloud Platform;Microsoft Azure;Raspberry Pi;Windows            0.000019
Android;Docker;MacOS;Microsoft Azure                                             0.000019
AWS;Docker;Kubernetes;MacOS;Raspberry Pi;Windows                                 0.000019
Name: Plat

## Groupby

* if you want specific results based on some specific columns then we need to group on that specific column using groupby 
function
* It involves splitting the object, applying the function and combining the results

In [17]:
df['Country'].value_counts()

United States       12469
India                8403
United Kingdom       3896
Germany              3890
Canada               2191
                    ...  
Marshall Islands        1
Gabon                   1
Kiribati                1
Saint Lucia             1
Nauru                   1
Name: Country, Length: 183, dtype: int64

In [18]:
#this object contains a bunch of groups
df.groupby(['Country'])

In [19]:
#create a variable of group of countries
country_group = df.groupby(['Country'])

In [20]:
#group by country name United States
country_group.get_group('United States')

MainBranch Hobbyist   Age  \
Respondent                                                                     
5           I used to be a developer by profession, but no...      Yes  31.0   
8                              I am a developer by profession      Yes  36.0   
14                             I am a developer by profession      Yes  27.0   
17                             I am a developer by profession      Yes  25.0   
18                             I am a developer by profession      Yes  32.0   
...                                                       ...      ...   ...   
58498                                                     NaN      Yes  32.0   
58896                                                     NaN      Yes  29.0   
60931                                                     NaN      Yes   NaN   
63077                                                     NaN      Yes   NaN   
64858                                                     NaN      Yes   NaN   

           Age1stCode CompFreq  CompTotal  ConvertedComp        Country  \
Respondent                                                                
5                  16      NaN        NaN            NaN  United States   
8                  12   Yearly   116000.0       116000.0  United States   
14                 13   Yearly    66000.0        66000.0  United States   
17                 14   Yearly    79000.0        79000.0  United States   
18                 12  Monthly   105000.0      1260000.0  United States   
...               ...      ...        ...            ...            ...   
58498              23      NaN        NaN            NaN  United States   
58896              10      NaN        NaN            NaN  United States   
60931              17      NaN        NaN            NaN  United States   
63077              20      NaN        NaN            NaN  United States   
64858              16      NaN        NaN            NaN  United States   

                    CurrencyDesc CurrencySymbol  ...  \
Respondent                                       ...   
5                            NaN            NaN  ...   
8           United States dollar            USD  ...   
14          United States dollar            USD  ...   
17          United States dollar            USD  ...   
18          United States dollar            USD  ...   
...                          ...            ...  ...   
58498                        NaN            NaN  ...   
58896                        NaN            NaN  ...   
60931                        NaN            NaN  ...   
63077                        NaN            NaN  ...   
64858                        NaN            NaN  ...   

                            SurveyEase           SurveyLength Trans  \
Respondent                                                            
5                                 Easy              Too short    No   
8                                 Easy  Appropriate in length    No   
14          Neither easy nor difficult  Appropriate in length    No   
17                                Easy  Appropriate in length    No   
18          Neither easy nor difficult  Appropriate in length    No   
...                                ...                    ...   ...   
58498       Neither easy nor difficult  Appropriate in length    No   
58896       Neither easy nor difficult  Appropriate in length    No   
60931       Neither easy nor difficult               Too long    No   
63077                              NaN                    NaN   NaN   
64858                              NaN                    NaN   NaN   

                                               UndergradMajor  \
Respondent                                                      
5           Computer science, computer engineering, or sof...   
8           Computer science, computer engineering, or sof...   
14          Computer science, computer engineering, or sof...   
17          Computer science, computer engineering, or s

In [21]:
#group by country name india
country_group.get_group('India')

MainBranch Hobbyist   Age Age1stCode  \
Respondent                                                                     
7                   I am a developer by profession      Yes   NaN         18   
22                  I am a developer by profession      Yes   NaN         18   
58                  I am a developer by profession      Yes   NaN         22   
63          I am a student who is learning to code      Yes  21.0         17   
149                 I am a developer by profession      Yes  36.0         31   
...                                            ...      ...   ...        ...   
54757                                          NaN      Yes   NaN         18   
55407                                          NaN      Yes   NaN        NaN   
62464                                          NaN      Yes   NaN        NaN   
62954                                          NaN      Yes   NaN        NaN   
64236                                          NaN      Yes   NaN        NaN   

           CompFreq   CompTotal  ConvertedComp Country          CurrencyDesc  \
Respondent                                                                     
7           Monthly         NaN            NaN   India  United States dollar   
22          Monthly         NaN            NaN   India          Indian rupee   
58              NaN         NaN            NaN   India          Indian rupee   
63              NaN         NaN            NaN   India                   NaN   
149          Yearly  21000000.0       293196.0   India          Indian rupee   
...             ...         ...            ...     ...                   ...   
54757           NaN         NaN            NaN   India                   NaN   
55407           NaN         NaN            NaN   India                   NaN   
62464           NaN         NaN            NaN   India                   NaN   
62954           NaN         NaN            NaN   India                   NaN   
64236           NaN         NaN            NaN   India                   NaN   

           CurrencySymbol  ...                  SurveyEase  \
Respondent                 ...                               
7                     USD  ...                         NaN   
22                    INR  ...                        Easy   
58                    INR  ...  Neither easy nor difficult   
63                    NaN  ...                        Easy   
149                   INR  ...  Neither easy nor difficult   
...                   ...  ...                         ...   
54757                 NaN  ...                        Easy   
55407                 NaN  ...                         NaN   
62464                 NaN  ...                         NaN   
62954                 NaN  ...                         NaN   
64236                 NaN  ...                         NaN   

                     SurveyLength Trans  \
Respondent                                
7                             NaN   NaN   
22          Appropriate in length   NaN   
58                       Too long   NaN   
63          Appropriate in length    No   
149         Appropriate in length    No   
...                           ...   ...   
54757       Appropriate in length    No   
55407                         NaN   NaN   
62464                         NaN   NaN   
62954                         NaN   NaN   
64236                         NaN   NaN   

                                               UndergradMajor  \
Respondent                                                      
7           Computer science, computer engineering, or sof...   
22          Computer science, computer engineering, or sof...   
58                              Web development or web design   
63                                                        NaN   
149         Computer science, computer engineering, or sof...   
...                                                       ...   
54757       Computer science, computer engineering, or sof...   
55407 

In [22]:
#using another way to group 
filt = df['Country'] == 'India'
df.loc[filt]

MainBranch Hobbyist   Age Age1stCode  \
Respondent                                                                     
7                   I am a developer by profession      Yes   NaN         18   
22                  I am a developer by profession      Yes   NaN         18   
58                  I am a developer by profession      Yes   NaN         22   
63          I am a student who is learning to code      Yes  21.0         17   
149                 I am a developer by profession      Yes  36.0         31   
...                                            ...      ...   ...        ...   
54757                                          NaN      Yes   NaN         18   
55407                                          NaN      Yes   NaN        NaN   
62464                                          NaN      Yes   NaN        NaN   
62954                                          NaN      Yes   NaN        NaN   
64236                                          NaN      Yes   NaN        NaN   

           CompFreq   CompTotal  ConvertedComp Country          CurrencyDesc  \
Respondent                                                                     
7           Monthly         NaN            NaN   India  United States dollar   
22          Monthly         NaN            NaN   India          Indian rupee   
58              NaN         NaN            NaN   India          Indian rupee   
63              NaN         NaN            NaN   India                   NaN   
149          Yearly  21000000.0       293196.0   India          Indian rupee   
...             ...         ...            ...     ...                   ...   
54757           NaN         NaN            NaN   India                   NaN   
55407           NaN         NaN            NaN   India                   NaN   
62464           NaN         NaN            NaN   India                   NaN   
62954           NaN         NaN            NaN   India                   NaN   
64236           NaN         NaN            NaN   India                   NaN   

           CurrencySymbol  ...                  SurveyEase  \
Respondent                 ...                               
7                     USD  ...                         NaN   
22                    INR  ...                        Easy   
58                    INR  ...  Neither easy nor difficult   
63                    NaN  ...                        Easy   
149                   INR  ...  Neither easy nor difficult   
...                   ...  ...                         ...   
54757                 NaN  ...                        Easy   
55407                 NaN  ...                         NaN   
62464                 NaN  ...                         NaN   
62954                 NaN  ...                         NaN   
64236                 NaN  ...                         NaN   

                     SurveyLength Trans  \
Respondent                                
7                             NaN   NaN   
22          Appropriate in length   NaN   
58                       Too long   NaN   
63          Appropriate in length    No   
149         Appropriate in length    No   
...                           ...   ...   
54757       Appropriate in length    No   
55407                         NaN   NaN   
62464                         NaN   NaN   
62954                         NaN   NaN   
64236                         NaN   NaN   

                                               UndergradMajor  \
Respondent                                                      
7           Computer science, computer engineering, or sof...   
22          Computer science, computer engineering, or sof...   
58                              Web development or web design   
63                                                        NaN   
149         Computer science, computer engineering, or sof...   
...                                                       ...   
54757       Computer science, computer engineering, or sof...   
55407 

In [23]:
df.loc[filt]['PlatformWorkedWith'].value_counts()

Windows                                                                                            628
Android                                                                                            384
Linux;Windows                                                                                      336
Linux                                                                                              301
Android;Windows                                                                                    123
                                                                                                  ... 
iOS;Linux;MacOS;WordPress                                                                            1
Android;Arduino;Google Cloud Platform;Linux;Windows                                                  1
AWS;Docker;Linux;MacOS;Slack Apps and Integrations;WordPress                                         1
Android;AWS;Docker;Google Cloud Platform;iOS;Kubernetes;Linux;MacOS;Slack

In [24]:
country_group['PlatformWorkedWith'].value_counts().head(50)

Country      PlatformWorkedWith                                                                                                                                                       
Afghanistan  Android                                                                                                                                                                      7
             Windows                                                                                                                                                                      5
             Android;Windows                                                                                                                                                              4
             Android;Linux;Windows;WordPress                                                                                                                                              3
             Android;Google Cloud Platform;Windows               

In [25]:
#get the percentage of popular platform in india
country_group['PlatformWorkedWith'].value_counts(normalize=True).loc['India']

PlatformWorkedWith
Windows                                  0.100496
Android                                  0.061450
Linux;Windows                            0.053769
Linux                                    0.048168
Android;Windows                          0.019683
                                           ...   
iOS;MacOS;Raspberry Pi;Windows           0.000160
iOS;MacOS;Slack Apps and Integrations    0.000160
iOS;MacOS;WordPress                      0.000160
iOS;Microsoft Azure                      0.000160
iOS;Raspberry Pi                         0.000160
Name: PlatformWorkedWith, Length: 1436, dtype: float64

In [26]:
#get the percentage of popular platform in United States
country_group['PlatformWorkedWith'].value_counts(normalize=True).loc['United States']

PlatformWorkedWith
Windows                                                    0.077134
Linux;Windows                                              0.034628
Linux                                                      0.032796
Microsoft Azure;Windows                                    0.021711
Linux;MacOS                                                0.012276
                                                             ...   
iOS;MacOS;Windows;WordPress                                0.000092
iOS;Microsoft Azure;Slack Apps and Integrations;Windows    0.000092
iOS;Microsoft Azure;Windows;WordPress                      0.000092
iOS;Raspberry Pi;Slack Apps and Integrations               0.000092
iOS;Slack Apps and Integrations                            0.000092
Name: PlatformWorkedWith, Length: 2607, dtype: float64

In [27]:
#median salary of country germany
country_group['ConvertedComp'].median().loc['Germany']

62697.0

In [28]:
#use agg() method and pass all the aggregated functions you want to use
country_group['ConvertedComp'].agg(['median', 'mean'])

median           mean
Country                                                     
Afghanistan                           15163.5  148290.125000
Albania                               15900.0   25611.000000
Algeria                                9432.0   10362.812500
Andorra                               88640.0   88640.000000
Angola                                 5292.0    5292.000000
...                                       ...            ...
Venezuela, Bolivarian Republic of...   3600.0    6280.611111
Viet Nam                              10344.0   28342.605769
Yemen                                 36000.0   36000.000000
Zambia                                 5452.0   17506.400000
Zimbabwe                               7200.0   26338.181818

[183 rows x 2 columns]

In [29]:
country_group['ConvertedComp'].median().loc['Canada']

68068.0

In [30]:
country_group['ConvertedComp'].agg(['median', 'mean']).loc['Canada']

median     68068.000000
mean      115101.775465
Name: Canada, dtype: float64

In [31]:
#get the details about country canada
filt = df['Country'] == 'Canada'
df.loc[filt]

MainBranch Hobbyist   Age Age1stCode CompFreq  \
Respondent                                                                      
26          I am a developer by profession       No   NaN         12   Yearly   
32          I am a developer by profession      Yes  31.0         12  Monthly   
51          I am a developer by profession       No   NaN         10   Yearly   
66          I am a developer by profession      Yes  21.0          9      NaN   
75          I am a developer by profession       No   NaN        NaN      NaN   
...                                    ...      ...   ...        ...      ...   
19609                                  NaN      Yes   NaN         10      NaN   
30010                                  NaN      Yes  28.0         16      NaN   
35616                                  NaN      Yes   NaN        NaN      NaN   
36174                                  NaN      Yes   NaN        NaN      NaN   
62834                                  NaN      Yes  17.0         10      NaN   

            CompTotal  ConvertedComp Country     CurrencyDesc CurrencySymbol  \
Respondent                                                                     
26            94000.0        71093.0  Canada  Canadian dollar            CAD   
32             3500.0        31764.0  Canada  Canadian dollar            CAD   
51           137000.0       103615.0  Canada  Canadian dollar            CAD   
66                NaN            NaN  Canada              NaN            NaN   
75                NaN            NaN  Canada  Canadian dollar            CAD   
...               ...            ...     ...              ...            ...   
19609             NaN            NaN  Canada              NaN            NaN   
30010             NaN            NaN  Canada              NaN            NaN   
35616             NaN            NaN  Canada              NaN            NaN   
36174             NaN            NaN  Canada              NaN            NaN   
62834             NaN            NaN  Canada              NaN            NaN   

            ...                  SurveyEase           SurveyLength Trans  \
Respondent  ...                                                            
26          ...                        Easy               Too long    No   
32          ...  Neither easy nor difficult               Too long    No   
51          ...                   Difficult               Too long   NaN   
66          ...  Neither easy nor difficult  Appropriate in length    No   
75          ...  Neither easy nor difficult  Appropriate in length   NaN   
...         ...                         ...                    ...   ...   
19609       ...  Neither easy nor difficult  Appropriate in length    No   
30010       ...                        Easy  Appropriate in length    No   
35616       ...                         NaN                    NaN   NaN   
36174       ...                         NaN                    NaN   NaN   
62834       ...                        Easy  Appropriate in length    No   

                                               UndergradMajor  \
Respondent                                                      
26          Computer science, computer engineering, or sof...   
32          Computer science, computer engineering, or sof...   
51                                   I never declared a major   
66          Computer science, computer engineering, or sof...   
75          Computer science, computer engineering, or sof...   
...                                                       ...   
19609       Computer science, computer engineering, or sof...   
30010       Computer science, computer engineering, or sof...   
35616                                                     NaN   
36174                                                     NaN   
62834                                                     NaN   

           WebframeDesireNextYear                  WebframeWorkedWith  \
Respondent                   

In [32]:
#get Windows users of canada
df.loc[filt]['PlatformWorkedWith'].str.contains('Windows')

Respondent
26       False
32        True
51       False
66       False
75       False
         ...  
19609    False
30010    False
35616      NaN
36174     True
62834     True
Name: PlatformWorkedWith, Length: 2191, dtype: object

In [33]:
#get total number of windows users of canada
df.loc[filt]['PlatformWorkedWith'].str.contains('Windows').sum()

988

In [34]:
#alternatively
country_group['PlatformWorkedWith'].apply(lambda x: x.str.contains('Windows').sum())

Country
Afghanistan                              24
Albania                                  23
Algeria                                  46
Andorra                                   2
Angola                                    2
                                       ... 
Venezuela, Bolivarian Republic of...     42
Viet Nam                                156
Yemen                                     6
Zambia                                    8
Zimbabwe                                 15
Name: PlatformWorkedWith, Length: 183, dtype: int64

### Sum of people from each country who use Python

In [35]:
country_group['LanguageWorkedWith'].apply(lambda x: x.str.contains('Python').sum())

Country
Afghanistan                              11
Albania                                  13
Algeria                                  40
Andorra                                   3
Angola                                    1
                                       ... 
Venezuela, Bolivarian Republic of...     29
Viet Nam                                102
Yemen                                     1
Zambia                                    4
Zimbabwe                                 13
Name: LanguageWorkedWith, Length: 183, dtype: int64

### Percentage of people from each country who use Python

In [36]:
#total number of respondents from each country
country_respondents = df['Country'].value_counts()
country_respondents

United States       12469
India                8403
United Kingdom       3896
Germany              3890
Canada               2191
                    ...  
Marshall Islands        1
Gabon                   1
Kiribati                1
Saint Lucia             1
Nauru                   1
Name: Country, Length: 183, dtype: int64

In [37]:
#total number of people from each country who knows python
country_uses_python = country_group['LanguageWorkedWith'].apply(lambda x: x.str.contains('Python').sum())
country_uses_python

Country
Afghanistan                              11
Albania                                  13
Algeria                                  40
Andorra                                   3
Angola                                    1
                                       ... 
Venezuela, Bolivarian Republic of...     29
Viet Nam                                102
Yemen                                     1
Zambia                                    4
Zimbabwe                                 13
Name: LanguageWorkedWith, Length: 183, dtype: int64

In [38]:
#create dataframe of count of people from that country and count of people who worked with python
python_df = pd.concat([country_respondents, country_uses_python], axis='columns', sort=False)
python_df

Country  LanguageWorkedWith
United States       12469                5964
India                8403                2670
United Kingdom       3896                1621
Germany              3890                1712
Canada               2191                1011
...                   ...                 ...
Marshall Islands        1                   0
Gabon                   1                   1
Kiribati                1                   0
Saint Lucia             1                   1
Nauru                   1                   0

[183 rows x 2 columns]

Here, 12469 counts of people are from united states and out of that 5964 counts of people works with python

In [39]:
#update meaningful name of columns of python_df
python_df.rename(columns = {'Country': 'NumRespondents', 'LanguageWorkedWith': 'NumKnowsPython'}, inplace=True)
python_df

NumRespondents  NumKnowsPython
United States              12469            5964
India                       8403            2670
United Kingdom              3896            1621
Germany                     3890            1712
Canada                      2191            1011
...                          ...             ...
Marshall Islands               1               0
Gabon                          1               1
Kiribati                       1               0
Saint Lucia                    1               1
Nauru                          1               0

[183 rows x 2 columns]

In [40]:
#create a python percentage column  
python_df['PercentKnowsPython'] = (python_df['NumKnowsPython']/python_df['NumRespondents']) * 100

In [41]:
python_df

NumRespondents  NumKnowsPython  PercentKnowsPython
United States              12469            5964           47.830620
India                       8403            2670           31.774366
United Kingdom              3896            1621           41.606776
Germany                     3890            1712           44.010283
Canada                      2191            1011           46.143314
...                          ...             ...                 ...
Marshall Islands               1               0            0.000000
Gabon                          1               1          100.000000
Kiribati                       1               0            0.000000
Saint Lucia                    1               1          100.000000
Nauru                          1               0            0.000000

[183 rows x 3 columns]

In [42]:
#sort countries by largest percentage of respondent who knows python
python_df.sort_values(by='PercentKnowsPython', ascending=False, inplace=True)
python_df

NumRespondents  NumKnowsPython  \
Saint Lucia                                      1               1   
Gabon                                            1               1   
Micronesia, Federated States of...               1               1   
Montenegro                                      13               9   
Brunei Darussalam                                3               2   
...                                            ...             ...   
San Marino                                       2               0   
Burundi                                          2               0   
Cape Verde                                       2               0   
Niger                                            5               0   
Nauru                                            1               0   

                                    PercentKnowsPython  
Saint Lucia                                 100.000000  
Gabon                                       100.000000  
Micronesia, Federated States of...          100.000000  
Montenegro                                   69.230769  
Brunei Darussalam                            66.666667  
...                                                ...  
San Marino                                    0.000000  
Burundi                                       0.000000  
Cape Verde                                    0.000000  
Niger                                         0.000000  
Nauru                                         0.000000  

[183 rows x 3 columns]

In [43]:
#get the statistics of United States
python_df.loc['United States']

NumRespondents        12469.00000
NumKnowsPython         5964.00000
PercentKnowsPython       47.83062
Name: United States, dtype: float64

In [44]:
#statistics of india
python_df.loc['India']

NumRespondents        8403.000000
NumKnowsPython        2670.000000
PercentKnowsPython      31.774366
Name: India, dtype: float64